src/backend/wayland/data.rs

```diff
@@ -0,0 +1,556 @@
+                    source,
+                } => {
+                    if let Some(window) = self.focus.pointer {
+                        self.record_event(WaylandEvent::Input(InputEvent::PointerAxis {
+                            event: WaylandPointerAxisEvent {
+                                device: self.pointer.unwrap(),
+                                vertical,
+                                horizontal,
+                                source,
+                                time,
+                                window,
+                            },
+                        }));
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl KeyboardHandler for WaylandBackendData {
+    fn enter(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _keyboard: &wl_keyboard::WlKeyboard,
+        surface: &wl_surface::WlSurface,
+        _serial: u32,
+        _raw: &[u32],
+        _keysyms: &[u32],
+    ) {
+        if let Some(window) = self.window_from_wl_surface(surface) {
+            self.focus.keyboard = Some(window.id);
+        }
+
+        // TODO: Held keys?
+    }
+
+    fn leave(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _keyboard: &wl_keyboard::WlKeyboard,
+        _surface: &wl_surface::WlSurface,
+        _serial: u32,
+    ) {
+        self.focus.keyboard.take();
+    }
+
+    fn press_key(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _keyboard: &wl_keyboard::WlKeyboard,
+        _serial: u32,
+        event: KeyEvent,
+    ) {
+        self.record_event(WaylandEvent::Input(InputEvent::Keyboard {
+            event: WaylandKeyboardKeyEvent {
+                key_code: event.raw_code,
+                state: KeyState::Pressed,
+                device: self.keyboard.unwrap(),
+                // TODO: Count
+                count: 0,
+                time: event.time,
+                window: self.focus.keyboard.unwrap(),
+            },
+        }));
+    }
+
+    fn release_key(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _keyboard: &wl_keyboard::WlKeyboard,
+        _serial: u32,
+        event: KeyEvent,
+    ) {
+        self.record_event(WaylandEvent::Input(InputEvent::Keyboard {
+            event: WaylandKeyboardKeyEvent {
+                key_code: event.raw_code,
+                state: KeyState::Released,
+                device: self.keyboard.unwrap(),
+                // TODO: Count
+                count: 0,
+                time: event.time,
+                window: self.focus.keyboard.unwrap(),
+            },
+        }));
+    }
+
+    fn update_modifiers(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _keyboard: &wl_keyboard::WlKeyboard,
+        _serial: u32,
+        _modifiers: Modifiers,
+    ) {
+        // TODO: Modifiers?
```

Comment:
I'm having trouble testing it, but the modifiers used by Smithay will have to be updated when the host compositor sends `wl_keyboard::modifers`, or it will have the same issue as https://github.com/Smithay/smithay/issues/646 when then window gains/loses focus with modifers active.

Reply:
https://docs.rs/xkbcommon/0.5.0/xkbcommon/xkb/struct.State.html#method.update_key

> This entry point is intended for programs which track the keyboard state explictly (like an evdev client). If the state is serialized to you by a master process (like a Wayland compositor) using functions like xkb_state_serialize_mods(), you should use xkb_state_update_mask() instead. The two functins should not generally be used together.

Perhaps smithay should have a way to use `update_mask` instead? This is also relevant to Cosmic's xdg-shell-wrapper which is used to implement things like the panel.

Reply:
Or if the compositor using the wayland backend wants to use its own keymap/group, maybe it should use the keys passed to `enter`. Don't know if it would be sensible to synthesize press/release events from that.